dm/syncer/syncer.go

```diff
@@ -2344,6 +2344,35 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 			if err2 != nil {
 				return err2
 			}
+		case *replication.TransactionPayloadEvent:
+			for _, tpev := range ev.Events {
+				switch tpevt := tpev.Event.(type) {
+				case *replication.RowsEvent:
+					eventIndex++
+					s.metricsProxies.Metrics.BinlogEventRowHistogram.Observe(float64(len(tpevt.Rows)))
+					ec.header.EventType = replication.WRITE_ROWS_EVENTv2 // FIXME
```

Comment:
yes `handleRowsEvent` will use EventType to know it's a INSERT or DELETE or UPDATE. we can use the header of `tpevt`?

Reply:
Looks like it is not exported:
https://github.com/go-mysql-org/go-mysql/blob/f8546806eee3bcb9692350f670bd456d8bf1537c/replication/row_event.go#L833

```
> github.com/pingcap/tiflow/dm/syncer.(*Syncer).Run() ./dm/syncer/syncer.go:2352 (hits goroutine(543):1 total:1) (PC: 0x65d4ee4)
  2347:				}
  2348:			case *replication.TransactionPayloadEvent:
  2349:				for _, tpev := range ev.Events {
  2350:					switch tpevt := tpev.Event.(type) {
  2351:					case *replication.RowsEvent:
=>2352:						eventIndex++
  2353:						s.metricsProxies.Metrics.BinlogEventRowHistogram.Observe(float64(len(tpevt.Rows)))
  2354:						ec.header.EventType = replication.WRITE_ROWS_EVENTv2 // FIXME
  2355:						sourceTable, err2 = s.handleRowsEvent(tpevt, ec)
  2356:						if sourceTable != nil && err2 == nil && s.cfg.EnableGTID {
  2357:							if _, ok := affectedSourceTables[sourceTable.Schema]; !ok {
(dlv) p tpevt
*github.com/go-mysql-org/go-mysql/replication.RowsEvent {
	Version: 2,
	tableIDSize: 6,
	tables: map[uint64]*github.com/go-mysql-org/go-mysql/replication.TableMapEvent [
		124: *(*"github.com/go-mysql-org/go-mysql/replication.TableMapEvent")(0xc000eafc00), 
	],
	needBitmap2: false,
	eventType: WRITE_ROWS_EVENTv2 (30),
	Table: *github.com/go-mysql-org/go-mysql/replication.TableMapEvent {
		flavor: "",
		tableIDSize: 6,
		TableID: 124,
		Flags: 1,
		Schema: []uint8 len: 4, cap: 999901, [116,101,115,116],
		Table: []uint8 len: 2, cap: 999895, [116,49],
		ColumnCount: 1,
		ColumnType: []uint8 len: 1, cap: 999891, [3],
		ColumnMeta: []uint16 len: 1, cap: 1, [0],
		NullBitmap: []uint8 len: 1, cap: 999889, [0],
		SignednessBitmap: []uint8 len: 1, cap: 999886, [0],
		DefaultCharset: []uint64 len: 0, cap: 0, nil,
		ColumnCharset: []uint64 len: 0, cap: 0, nil,
		SetStrValue: [][][]uint8 len: 0, cap: 0, nil,
		setStrValueString: [][]string len: 0, cap: 0, nil,
		EnumStrValue: [][][]uint8 len: 0, cap: 0, nil,
		enumStrValueString: [][]string len: 0, cap: 0, nil,
		ColumnName: [][]uint8 len: 0, cap: 0, nil,
		columnNameString: []string len: 0, cap: 0, nil,
		GeometryType: []uint64 len: 0, cap: 0, nil,
		PrimaryKey: []uint64 len: 0, cap: 0, nil,
		PrimaryKeyPrefix: []uint64 len: 0, cap: 0, nil,
		EnumSetDefaultCharset: []uint64 len: 0, cap: 0, nil,
		EnumSetColumnCharset: []uint64 len: 0, cap: 0, nil,},
	TableID: 124,
	Flags: 1,
	ExtraData: []uint8 len: 0, cap: 999856, [],
	ColumnCount: 1,
	ColumnBitmap1: []uint8 len: 1, cap: 999855, [255],
	ColumnBitmap2: []uint8 len: 0, cap: 0, nil,
	Rows: [][]interface {} len: 1, cap: 1, [
		[
			*(*interface {})(0xc001fce4a0),
		],
	],
	SkippedColumns: [][]int len: 1, cap: 1, [
		[],
	],
	parseTime: false,
	timestampStringLocation: *time.Location nil,
	useDecimal: false,
	ignoreJSONDecodeErr: false,}
(dlv) p tpevt.eventType
WRITE_ROWS_EVENTv2 (30)
```

Should I create a PR for go-mysql to export it? or is there another way?

Reply:
Sorry I should say `tpev.Header.EventType`

Reply:
```
(dlv) list
> github.com/pingcap/tiflow/dm/syncer.(*Syncer).Run() ./dm/syncer/syncer.go:2354 (hits goroutine(2074):1 total:1) (PC: 0x65d4f53)
  2349:				for _, tpev := range ev.Events {
  2350:					switch tpevt := tpev.Event.(type) {
  2351:					case *replication.RowsEvent:
  2352:						eventIndex++
  2353:						s.metricsProxies.Metrics.BinlogEventRowHistogram.Observe(float64(len(tpevt.Rows)))
=>2354:						ec.header.EventType = tpev.Header.EventType
  2355:						sourceTable, err2 = s.handleRowsEvent(tpevt, ec)
  2356:						if sourceTable != nil && err2 == nil && s.cfg.EnableGTID {
  2357:							if _, ok := affectedSourceTables[sourceTable.Schema]; !ok {
  2358:								affectedSourceTables[sourceTable.Schema] = make(map[string]struct{})
  2359:							}
(dlv) p tpev.Header.EventType
WRITE_ROWS_EVENTv2 (30)
```

Reply:
4554b518e4969fb7654ec2a172264403eabff8dd